bluesky/callbacks/waterfall.py

```diff
@@ -0,0 +1,168 @@
+        # flag to prevent update
+        self.halt = False
+        # add sliders, which store information
+        y_offset_slider_ax = self.fig.add_axes([0.15, 0.95, 0.3, 0.035])
+        self.y_offset_slider = Slider(y_offset_slider_ax,
+                                      'y-offset', 0.0, 1.0,
+                                      valinit=0.1, valfmt='%1.2f')
+        self.y_offset_slider.on_changed(self.update_y_offset)
+
+        x_offset_slider_ax = self.fig.add_axes([0.6, 0.95, 0.3, 0.035])
+        self.x_offset_slider = Slider(x_offset_slider_ax,
+                                      'x-offset', 0.0, 1.0,
+                                      valinit=0.1, valfmt='%1.2f')
+        self.x_offset_slider.on_changed(self.update_x_offset)
+        # init
+        self.update(self.key_list, self.int_data_list, refresh=True)
+
+    def update(self, key_list=None, int_data_list=None, refresh=False):
+        """top method to update information carried by class and plot
+
+        Parameters
+        ----------
+        key_list : list, optional
+            list of keys. default to None.
+        int_data_list : list, optional
+            list of 1D data. default to None.
+        refresh : bool, optional
+            option to set refresh or not. default to False.
+        """
+        if not int_data_list:
+            print("INFO: no reduced data was fed in, "
+                  "waterfall plot can't be updated")
+            self.halt = True
+            self.no_int_data_plot(self.ax, self.canvas)
+            return
+        # refresh list
+        if refresh:
+            self.key_list = []
+            self.int_data_list = []
+        self.key_list.extend(key_list)
+        self.int_data_list.extend(int_data_list)
+        self._adapt_data_list(self.int_data_list)
+        # generate plot
+        self.halt = False
+        self._update_plot()  # use current value of x,y offset
+
+    def _adapt_data_list(self, int_data_list):
+        """method to return statefull information of 1D data list"""
+        x_array_list = []
+        y_array_list = []
+        # parse
+        for x, y in int_data_list:
+            x_array_list.append(x)
+            y_array_list.append(y)
+        self.x_array_list = x_array_list
+        self.y_array_list = y_array_list
+        # stateful information
+        self.y_dist = np.max(y_array_list) - np.min(y_array_list)
+        self.x_dist = np.max(x_array_list) - np.min(x_array_list)
+
+    def on_plot_hover(self, event):
+        """callback to show legend when click on one of curves"""
+        line = event.artist
+        name = line.get_label()
+        line.axes.legend([name], handlelength=0,
+                         handletextpad=0, fancybox=True)
+        line.figure.canvas.draw_idle()
+
+    def _update_plot(self, x_offset_val=None, y_offset_val=None):
+        """core method to update x-, y-offset sliders"""
+        self.ax.set_facecolor('w')
+        self.ax.cla()
```

Comment:
This is a very inefficient way to do this,  it is better to keep a mapping between the `Line2D` instances, the raw data they came from and the offset associated with that line and update as required.

Reply:
So it seems better to manage ``Line2D`` instance for each curve and use ``set_xdata, ``set_ydata`` methods to update the array and then ``draw``?

(found example might be related: https://matplotlib.org/gallery/api/line_with_text.html#sphx-glr-gallery-api-line-with-text-py) 

Also, under this context do we need to worry about cleaning ``Axes`` before draw?